### PR TITLE
[storage][pruner] Incrementally prune transaction accumulator

### PR DESCRIPTION
**Pull request by zcchahaha**
### Description
To address https://github.com/aptos-labs/aptos-core/issues/1288
### Test Plan
Added a UT.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2734)

### CODE REVIEW

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 1]
```diff
@@ -316,6 +317,73 @@ impl TransactionStore {
...
+        //    b (level position = 0)      c(level position = 1)
+        //
+        // To find the first ancestor which is a left child of its parent, we can keep diving the
+        // level position by 2 (to find the ancestor) until we get a level position which is a
+        // multiple of 2 (to make sure the ancestor is a left child of its parent).
+        while version & 1 != 0 {
+            version >>= 1;
+            first_ancestor_that_is_a_left_child_level += 1;
+        }
+        Position::from_level_and_pos(first_ancestor_that_is_a_left_child_level, version)
```
Comment (msmouse) [inline storage/aptosdb/src/transaction_store/mod.rs:384]: This is not right..

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 2]
```diff
@@ -316,6 +317,73 @@ impl TransactionStore {
...
+        //         /                                \
+        //    b (level position = 0)      c(level position = 1)
+        //
+        // To find the first ancestor which is a left child of its parent, we can keep diving the
+        // level position by 2 (to find the ancestor) until we get a level position which is a
+        // multiple of 2 (to make sure the ancestor is a left child of its parent).
+        while version & 1 != 0 {
+            version >>= 1;
+            first_ancestor_that_is_a_left_child_level += 1;
+        }
```
Comment (msmouse) [inline storage/aptosdb/src/transaction_store/mod.rs:383]: beautiful, but
`first_ancestor_that_is_a_left_child_level = version.trailing_ones()`

File: storage/aptosdb/src/transaction_store/mod.rs (hunk 1 above)
Comment (msmouse) [reply storage/aptosdb/src/transaction_store/mod.rs:384]: oh the name is not right..

File: storage/aptosdb/src/pruner/transaction_store/test.rs [hunk 3]
```diff
@@ -191,6 +193,25 @@ fn verify_txn_in_store(
...
+        Position::root_from_leaf_index(least_readable_version).left_child()
+    } else {
+        Position::root_from_leaf_index(least_readable_version)
+    };
+    let least_readable_position_postorder = least_readable_position.to_postorder_index();
+    for i in 0..least_readable_position_postorder {
+        assert!(ledger_store
+            .get(Position::from_postorder_index(i).unwrap())
+            .is_err())
+    }
```
[the remark this answers is not in the document]
Comment (zcchahaha) [reply storage/aptosdb/src/pruner/transaction_store/test.rs:223]: Done

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 4]
```diff
@@ -316,6 +317,73 @@ impl TransactionStore {
...
+                continue;
+            }
+
+            let mut first_ancestor_that_is_a_left_child =
+                self.find_first_ancestor_that_is_a_left_child(version_to_delete);
+
+            // This assertion is true because we skip the leaf nodes with address which is a
+            // a multiple of 2.
+            assert!(!first_ancestor_that_is_a_left_child.is_leaf());
+
```
[the remark this answers is not in the document]
Comment (zcchahaha) [reply storage/aptosdb/src/transaction_store/mod.rs:348]: Done.

File: storage/aptosdb/src/transaction_store/mod.rs (hunk 1 above)
Comment (zcchahaha) [reply storage/aptosdb/src/transaction_store/mod.rs]: Done

File: storage/aptosdb/src/transaction_store/mod.rs (hunk 2 above)
Comment (zcchahaha) [reply storage/aptosdb/src/transaction_store/mod.rs]: Done.

File: storage/aptosdb/src/transaction_store/mod.rs [hunk 5]
```diff
@@ -316,6 +317,65 @@ impl TransactionStore {
...
+    ///
+    /// To avoid always pruning a full left subtree, we uses the following algorithm.
+    /// For each leaf with an odd leaf index.
+    /// 1. From the bottom upwards, find the first ancestor that's a left child of its parent.
+    /// (the position of which can be got by popping "1"s from the right of the leaf address).
+    /// Note that this node DOES NOT become non-useful.
+    /// 2. From the node found from the previous step, delete both its children non-useful, and go
+    /// to the right child to repeat the process until we reach a leaf node.
+    /// More details are in this issue https://github.com/aptos-labs/aptos-core/issues/1288.
+    pub fn prune_transaction_accumulator(
```
Comment (grao1991) [inline storage/aptosdb/src/transaction_store/mod.rs:330]: I think we can also remove the get_min_proof_node function below which was used for the old algorithm?

Comment (zcchahaha) [reply storage/aptosdb/src/transaction_store/mod.rs:330]: Done

**Review by grao1991 — APPROVED** [review]

File: storage/aptosdb/src/pruner/transaction_store/test.rs [hunk 6]
```diff
@@ -132,6 +134,16 @@ fn verify_txn_store_pruner(
         );
         for j in 0..i {
             verify_txn_not_in_store(transaction_store, &txns, j as u64, ledger_version);
+            // Ensure that transaction accumulator is pruned in DB. This can be done by trying to
+            // read transaction proof.
+            // Note: we only prune versions which are odd numbers because the even versions will be
+            // pruned in the iteration of even_version + 1. So if the end version, i - 1, is an even
+            // version, it will not be pruned.
```
Comment (msmouse) [inline storage/aptosdb/src/pruner/transaction_store/test.rs:141]: 👍

File: storage/aptosdb/src/pruner/transaction_store/test.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/aptosdb/src/pruner/transaction_store/test.rs:223]: I mean let's call `db.ledger_store.get_transaction_info_with_proof()`?

[the remark this answers is not in the document]
Comment (msmouse) [reply storage/aptosdb/src/pruner/transaction_store/test.rs:223]: oh I see you call it elsewhere..

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/pruner/transaction_store/test.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zcchahaha) [reply storage/aptosdb/src/pruner/transaction_store/test.rs:223]: Maybe we can just remove this as we already have the check on line 142

**Comment by zcchahaha** [comment]
I disabled the flaky test for now and will reenable after changing the pruner threading model.

**Pull request merged**